### PR TITLE
Publish new versions to test PyPI on package version bump.

### PR DESCRIPTION
**Pull request by obi1kenobi**
When a PR updates the `pyproject.toml` version to a number that does not currently exist on PyPI (the main one, not test), a successful test run will cause the package to be built and published on test PyPI under that version by triggering the test release workflow.

We use caching to avoid spamming the PyPI metadata APIs on every PR, and instead only make one request per new version.

We have updated the release workflow to always overwrite test PyPI publishes, so I decided to not make the test publishes use `rc` versions to avoid introducing additional implementation complexity. If you feel that the extra complexity is worth it, I'm happy to switch to generating an `rc` release instead.

### CODE REVIEW

File: .github/workflows/_test.yml [hunk 1]
```diff
@@ -55,3 +55,32 @@ jobs:
...
           echo "$STATUS" | grep 'nothing to commit, working tree clean'
+
+      # If this PR is for a new, not-yet-published version,
+      # we want to run a build and prepare it for uploading to test PyPI.
+      - name: Check if version exists on PyPI
+        shell: bash
+        run: |
+          set -euxo pipefail
+
+          PKG_NAME="$(poetry version | cut -d' ' -f 1)"
```
Comment (baskaryan) [inline .github/workflows/_test.yml:66]: ubernit: should there be space after d before quote `-d ' '`

Comment (obi1kenobi) [reply .github/workflows/_test.yml:83]: Long time habit to never leave whitespace after the `-d` parameter to `cut` specifically. I'm not even sure why we always did that at Kensho... 😅 Thanks!

**Comment by obi1kenobi** [comment]
Needs #12668 to merge first plus a bit of configuration on test PyPI to configure trusted publishing.

**Comment by obi1kenobi** [comment]
If we're still interested in publishing to test PyPI when a PR changes the `pyproject.toml` version number, I think this PR should be good to go.

**Pull request closed without merging**